File: 61.rotate_list/rotateRight.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (head === null || head.next === null || k === 0) return head;
    var newhead;
    var fast = head;
    var slow = head;
    for(var i = 0; i < k; i++){
        if(fast.next === null){ //rotate back to head
            fast = head;
        }else{
            fast = fast.next;
        }
    }
    while(fast.next !== null){//slow now reaches the node to be shifted
        fast = fast.next;
        slow = slow.next;
    }
    fast.next = head;
    newhead = slow.next;
    slow.next = null;
    return newhead;
};

File: 61.rotate_list/rotateRight.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:

        p = head
        # get length
        l = 0
        while p:
            l += 1
            p = p.next
        if not l:
            return None
        k = k % l
        if not k or not head:
            return head
        fast = slow = head
        while k:
            fast = fast.next
            k -= 1
        while fast and fast.next:
            slow = slow.next
            fast = fast.next
        new_head = slow.next
        slow.next = None
        fast.next = head
        return new_head
